### ISSUE TITLE
MySQL 8.0 Group Replication / InnoDB Cluster support

### ISSUE THREAD

**Issue by camo-f** [docker-library/mysql#743]
Hello,

MySQL provides a way to create replicated databases through [InnoDB Cluster](https://dev.mysql.com/doc/refman/8.0/en/mysql-innodb-cluster-introduction.html). It requires the plugin [Group Replication](https://dev.mysql.com/doc/refman/8.0/en/group-replication.html) and can be managed with [MySQL Shell](https://dev.mysql.com/doc/refman/8.0/en/mysql-shell-userguide.html). 

I encountered two issues while trying to use InnoDB Cluster with `mysql:8.0` (up to 8.0.23).

___
MySQL Shell is not installed by default. This is debatable as it can be considered as a part of MySQL 8.0, but it would increase the size of the base image. Installation is quite easy though and could be done in the Dockerfile to enable availability on startup :
```
echo 'deb http://repo.mysql.com/apt/debian/ buster mysql-tools' >> /etc/apt/sources.list.d/mysql.list
apt install -y mysql-shell
```
On a side note, the fork made by the MySQL team ships with MySQL Shell.
___
Group Replication (the core plugin of InnoDB Cluster) can be configured with variables (starting with `group_replication*`). These variables cannot be read unless the plugin is loaded, but it cannot be done during initialization phase (this is a MySQL limitation).

In other words, it is not possible to provide a full configuration file with `group_replication` variables when starting a database from scratch. Variables must be added after initialization, making databases startup more cumbersome and harder to automate. 

A solution would be to ignore some configuration files or a whole directory during initialization (in this case, a file containing all `group_replication` variables), by using an environment variable for example.
___
Thanks

**Comment by camo-f** [comment]
Hi, any news on this ? Should I send a PR for it ?

**Comment by yosifkit** [comment]
As far as installing MySQL Shell by default, I would say no since it is used for managing a cluster, not by the cluster itself. It is also fairly large.

```console
root@82aac1ba71e5:/# apt install --no-install-recommends mysql-shell
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  libcurl4 libexpat1 libgssapi-krb5-2 libk5crypto3 libkeyutils1 libkrb5-3
  libkrb5support0 libmpdec2 libnghttp2-14 libpsl5 libpython3-stdlib libpython3.7
  libpython3.7-minimal libpython3.7-stdlib librtmp1 libssh2-1 mime-support python3
  python3-minimal python3.7 python3.7-minimal
Suggested packages:
  krb5-doc krb5-user python3-doc python3-tk python3-venv python3.7-venv
  python3.7-doc binutils binfmt-support
Recommended packages:
  ca-certificates krb5-locales publicsuffix bzip2 file
The following NEW packages will be installed:
  libcurl4 libexpat1 libgssapi-krb5-2 libk5crypto3 libkeyutils1 libkrb5-3
  libkrb5support0 libmpdec2 libnghttp2-14 libpsl5 libpython3-stdlib libpython3.7
  libpython3.7-minimal libpython3.7-stdlib librtmp1 libssh2-1 mime-support
  mysql-shell python3 python3-minimal python3.7 python3.7-minimal
0 upgraded, 22 newly installed, 0 to remove and 0 not upgraded.
Need to get 22.7 MB of archives.
After this operation, 125 MB of additional disk space will be used.
Do you want to continue? [Y/n] 
```

We could at least add the sources.list for it so that it is easy to install.

I'm slightly confused about needing to set any `group_replication` settings since if you are using MySQL Shell to administer, then: 
> AdminAPI removes the need to work directly with Group Replication in InnoDB Clusters
>
> \- https://dev.mysql.com/doc/mysql-shell/8.0/en/mysql-innodb-cluster.html

**Comment by camo-f** [comment]
@yosifkit I indeed overlooked the size of the package. Also maybe cluster management should be split up from MySQL and be an image/microservice by itself.  
I second adding the mysql.list, it will be enough to facilitate the use of MySQL Shell. 

Regarding `group_replication` settings, MySQL Shell is an additional layer to configure `group_replication` settings transparently.  The main goal is to configure most options before starting InnoDB Cluster to centralize MySQL server settings in a configuration directory for better management and versioning. This might be out of scope of the Docker image though.

**Comment by yosifkit** [comment]
Sorry for the long tail here. :bowing_man:

As far as a folder of configuration files (`*.cnf`), that's covered by the [docs](https://github.com/docker-library/docs/tree/a8cf45e9726be39bee12c9bfb229cc65160d855b/mysql#using-a-custom-mysql-configuration-file). Just drop them into `/etc/mysql/conf.d/` in either the Debian or Oracle Linux based image at runtime or in a custom image `FROM` them.

The new Oracle Linux based images contain mysql-shell by default: https://github.com/docker-library/mysql/blob/37981f652a98b8fc26f487be9eda167de4689d84/8.0/Dockerfile.oracle#L106-L111

Do we still need to add the `apt` sources line for users wanting to install it in a Debian-based image?

**Comment by camo-f** [comment]
Hi @yosifkit, thank you for the update.

I think adding `mysql-tools` to `sources.list` would be nice. This issue was initially on a Debian based image, and it will facilitate installing `mysql-shell` after starting the container.

On a side note, I'm curious about the Oracle Linux based image, but I did not find any documentation here or on Docker Hub about the differences with the Debian one. It seems new, do you have a quick explanation by any chance ?

Thanks a lot

**Comment by tianon** [comment]
See #680 -- the intention is to eventually replace the Debian variants by their Oracle counterparts (in the next major/minor release if we have to, but I'd love to be able to do so sooner, given that Oracle more officially supports MySQL on Oracle Linux for obvious reasons :smile:).

**Comment by tianon** [comment]
Unfortunately, I think that's probably the most we're going to do here in the image itself :bow:
